server/base/src/main/java/org/apache/accumulo/server/rpc/CustomNonBlockingServer.java

```diff
@@ -112,29 +112,20 @@ protected FrameBuffer createFrameBuffer(final TNonblockingTransport trans,
    * extract the client's network location before accepting the request.
    */
   private class CustomFrameBuffer extends FrameBuffer {
+    private final String clientAddress;
 
     public CustomFrameBuffer(TNonblockingTransport trans, SelectionKey selectionKey,
         AbstractSelectThread selectThread) throws TTransportException {
       super(trans, selectionKey, selectThread);
-    }
-
-    @Override
-    public void invoke() {
-      if (trans_ instanceof TNonblockingSocket) {
-        TNonblockingSocket tsock = (TNonblockingSocket) trans_;
-        Socket sock = tsock.getSocketChannel().socket();
-        TServerUtils.clientAddress
-            .set(sock.getInetAddress().getHostAddress() + ":" + sock.getPort());
-      }
-      super.invoke();
+      this.clientAddress = getClientAddress();
```

Comment:
I didn't look at it super long but from my testing it looked like invoke() and read() methods were not called on the same thread when i was logging the Thread ids. But also looking at it again another issue maybe be that read() is called before invoke() so the value isn't set yet. See:
https://github.com/apache/thrift/blob/9c0de2d1eb343910213c62325f73e3bb72361c22/lib/java/src/main/java/org/apache/thrift/server/AbstractNonblockingServer.java#L185-L196

For the constructor and frame buffer, I chose to put it there because looking at the source it appears to be a new frame buffer created for each connection which is why I moved it to the constructor.
https://github.com/apache/thrift/blob/9c0de2d1eb343910213c62325f73e3bb72361c22/lib/java/src/main/java/org/apache/thrift/server/TNonblockingServer.java#L213-L233

Reply:
Also, the FrameBuffer takes TNonblockingSocket as an argument for it's constructor and it's final in the buffer so the buffer can only be attached to a single client socket so it can't be re-used for a different connection. https://github.com/apache/thrift/blob/9c0de2d1eb343910213c62325f73e3bb72361c22/lib/java/src/main/java/org/apache/thrift/server/AbstractNonblockingServer.java#L286

Reply:
In regards to similar types and the instanceof check, the other location was done in ClientInfoProcessorFactory but those types are of TSocket which is not the same as TNonblockingSocket so you can't re-use the same code as the types are incompatible for finding the socket. On TSocket you need to call getSocket() and for TNonblockingSocket you need to call getSocketChannel().socket() (The joys/fun of blockin gvs nonblocking io)

https://github.com/apache/accumulo/blob/a541197cd3304910cf80dc83b7063a2bcd3eb1d6/server/base/src/main/java/org/apache/accumulo/server/rpc/ClientInfoProcessorFactory.java#L43-L55

There is only one implementation of TNonblockingTransport that I can see and that is TNonblockingSocket.

Reply:
Okay, if you checked that setting it in the constructor is sufficient, that's fine. I think it's still a problem that the ThreadLocal is no longer being set and we're still relying on it for the other places we read the client address. We need to either set it or remove it entirely, and extract the client address some other way for those other places where we get the client address.

Reply:
I'll go ahead take a look later today or tomorrow at where else it is being used and I'll see how to handle that. I will also double check that the constructor should be a safe place to set it. If we need to use it elsewhere i may try and figure out  something different than what I did but not sure yet.

Reply:
For what it's worth, if upstream thrift code provided easy access to the client address in its own API, then we wouldn't have to do any of this forking of thrift code into our project and modifying it to extract the client's address. It might be worth trying to figure out what we would prefer the Thrift API to look like to get the address (and to allow us to detect this weird case that we're logging), and to push it upstream, so we don't have to do it in our code at all, and all this could just be removed.